### PR TITLE
ENH: Add a mypy plugin for inferring platform-specific `np.number` precisions

### PR DESCRIPTION
**Pull request by BvB93**
Alternative to https://github.com/numpy/numpy/pull/17514.

This PR adds a mypy plugin that automatically assigns a platform-specific precision 
to certain `np.number` subclasses (or aliases thereof) while static type checking. 
Most notably this includes `np.int_`, which is always equivalent to `np.int32` on windows 
but `np.int64` on most (all?) other 64-bit platforms.

While https://github.com/numpy/numpy/pull/17514 attempts to accomplish this via a dynamically generated `.pyi` file, 
this PR instead uses a [mypy plugin](https://mypy.readthedocs.io/en/stable/extending_mypy.html). Note that while usage of the plugin is completelly 
optional, without it the precision of the likes of `np.int_` will be inferred as `Any`.

### CODE REVIEW

**Comment by BvB93** [comment]
Hm, anyone knows why only PyPy cannot seem to find the new `numpy.typing` module?

File: numpy/typing/mypy_plugin.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
+"""A module containing `numpy`-specific plugins for mypy."""
+
+from __future__ import annotations
```
Comment (BvB93) [inline numpy/typing/mypy_plugin.py:3]: To double check: We did drop support for python 3.6, right?

Comment (BvB93) [reply numpy/typing/mypy_plugin.py]: The `pypy37` tests are still running on python 3.6, so I guess that's a no.
Removed as of https://github.com/numpy/numpy/pull/17843/commits/f85cebade4a68b16da448f3b71285cf20b3d115d.

Comment (seberg) [reply numpy/typing/mypy_plugin.py]: We dropped it, if there are still tests running, that is probably a small error, but if its no difference here...

Comment (mattip) [reply numpy/typing/mypy_plugin.py]: whoops, [this is wrong](https://github.com/numpy/numpy/blob/master/.github/workflows/build_test.yml#L187), it should be 
```
wget -q https://downloads.python.org/pypy/pypy3.7-v7.3.2-linux64.tar.bz2 -O pypy.tar.bz2
```
I will submit a Pr to fix this.

File: numpy/typing/tests/data/reveal/arithmetic.py [hunk 2]
```diff
@@ -126,7 +126,7 @@
 reveal_type(c16 + b)  # E: numpy.complexfloating[numpy.typing._64Bit, numpy.typing._64Bit]
 reveal_type(c16 + c)  # E: numpy.complexfloating[numpy.typing._64Bit, numpy.typing._64Bit]
 reveal_type(c16 + f)  # E: numpy.complexfloating[numpy.typing._64Bit, numpy.typing._64Bit]
-reveal_type(c16 + i)  # E: numpy.complexfloating[Any, Any]
+reveal_type(c16 + i)  # E: numpy.complexfloating[numpy.typing._
```
Comment (rgommers) [inline numpy/typing/tests/data/reveal/arithmetic.py:129]: What's going on here? Looks like invalid syntax, but repeated below so probably on purpose.

**Comment by rgommers** [comment]
Interesting, tricky puzzle to get the details of this typing right. I'm not sure if one approach is clearly preferable to the other. 

The mypy plugin may possibly be slightly more robust to situations like cross-compiling (e.g. there's an open PR for ARM64 build on Windows, which may get the `intp` size wrong). However those are pretty niche cases that likely won't matter at all in practice. 

On the other hand, the plugin approach will require the user to add it in their own `mypy.ini` as well as have some runtime overhead (any idea how much?).

File: numpy/typing/tests/data/reveal/arithmetic.py (hunk 2 above)
Comment (BvB93) [reply numpy/typing/tests/data/reveal/arithmetic.py:129]: So the comment on the right contains the type as expected by mypy. The problem is that the value is platform-dependent while the comment is... just a normal static comment. The workaround I employed here is the use the fact that the `reveal` tests actually search for substring matches, _i.e._ it checkes whether `comment is in expected_value`.

Comment (rgommers) [reply numpy/typing/tests/data/reveal/arithmetic.py:129]: Makes sense, thanks for the explanation.

**Comment by BvB93** [comment]
> On the other hand, the plugin approach will require the user to add it in their own `mypy.ini` as well as have some runtime overhead (any idea how much?).

The runtime overhead should still be zero, as the new `numpy.typing.mypy_plugin` module should generally only be imported by mypy itself (so during static type checking). There are a couple of contingencies inside the module to ensure that it _can_ always be imported during runtime, but this is moreso because the numpy tests require so (regardless of what optional dependencies are present or not).

File: numpy/typing/tests/data/reveal/arithmetic.py (hunk 2 above)
Comment (BvB93) [reply numpy/typing/tests/data/reveal/arithmetic.py:129]: Xref https://github.com/numpy/numpy/issues/17617 as it is very much relevant for this particular issue.

Comment (eric-wieser) [reply numpy/typing/tests/data/reveal/arithmetic.py:129]: Perhaps worth a comment on the line above,
```suggestion
# note this comment is deliberate truncated as the result varies by platform
reveal_type(c16 + i)  # E: numpy.complexfloating[numpy.typing._
```

**Comment by rgommers** [comment]
> The runtime overhead should still be zero, as the new `numpy.typing.mypy_plugin` module should generally only be imported by mypy itself (so during static type checking).

Yeah I get that - I meant runtime of `mypy` (which can be slow).

**Comment by BvB93** [comment]
> Yeah I get that - I meant runtime of mypy (which can be slow).

Ah right, when running mypy on the script below I'm observing a minor increase in runtime when initially generating the mypy cache (~7.4 vs ~7.8 sec). Once the cache has been generated the difference disappears in its entirety though (~0.4 sec in both cases).

```python
import numpy as np
```

File: numpy/typing/tests/data/reveal/arithmetic.py (hunk 2 above)
Comment (BvB93) [reply numpy/typing/tests/data/reveal/arithmetic.py:132]: > Perhaps worth a comment on the line above,

Good call, done as of https://github.com/numpy/numpy/pull/17843/commits/b1b047e0b977463ef97e54bf5a9685b39ae3dd93.

**Comment by charris** [comment]
Needs rebase.

**Comment by BvB93** [comment]
> Needs rebase.

Yup, I'm planning on adding the release note + rebase later this evening.

**Comment by BvB93** [comment]
Alright, the branch has been rebased and a release note (https://github.com/numpy/numpy/pull/17843/commits/5c7c22abee421947133d0473adcc3d3a31d593cf) has been added.

**Comment by rgommers** [comment]
This accumulated another merge conflict unfortunately.

**Comment by BvB93** [comment]
Give me a sec.

**Comment by BvB93** [comment]
And the conflict has been resolved.

**Review by rgommers — APPROVED** [review]
Thanks @BvB93, in it goes!

**Pull request merged**